### PR TITLE
KL-UCB

### PR DESCRIPTION
**Pull request by Arkadiy-Vladimirov**
The PR proposes `KL-UCB` recommender based on the [article by Aurélien Garivier and Olivier Cappé](https://arxiv.org/abs/1102.2490). As a recommender model `KL-UCB` operates just as the regular `UCB` with the only difference in upper bound computation. This allowed to  implement `KL-UCB` as a subclass of `UCB` with basically the only `_calc_item_popularity` method overridden - one that calculates the bound. `__init__` method is slightly changed only to provide more adequate default parameter `exploration_coef = 0` as it was claimed to be the most suitable in practice by the authors of the [article](https://arxiv.org/abs/1102.2490). You may read more information on `KL-UCB` when recompile the documentation or in the class docstring directly. Evaluation is done in `02_model_comparison.ipynb` notebook. You may take a look at metrics there or in the picture below.

KL-UCB_metrics

### CODE REVIEW

File: replay/models/kl_ucb.py [hunk 1]
```diff
@@ -0,0 +1,154 @@
+import math
+
+from typing import Optional
+from replay.models import UCB
+from pyspark.sql.types import DoubleType
+from pyspark.sql.functions import udf
+from scipy.optimize import root_scalar
+
+
+class KL_UCB(UCB):
```
Comment (shashist) [inline replay/models/kl_ucb.py:10]: Thank you for the thorough documentation. What do you mean by `Single actor`?

File: replay/models/kl_ucb.py [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+            Could be changed after model training by setting the `sample`
+            attribute.
+        :param seed: random seed. Provides reproducibility if fixed
+        """
+
+        super().__init__(exploration_coef, sample, seed)
+
+    def _calc_item_popularity(self):
+
+        rhs = math.log(self.full_count) \
```
Comment (shashist) [inline replay/models/kl_ucb.py:116]: Could you make this variable name longer, e.g. right_bound?

**Comment by shashist** [comment]
`KL_UCB` uses log with both positive and negative relevance. Please check `fit_predict_add_res()` in notebook to use pos_neg_train for `KL_UCB`

**Comment by Arkadiy-Vladimirov** [comment]
> `KL_UCB` uses log with both positive and negative relevance. Please check `fit_predict_add_res()` in notebook to use `pos_neg_train` for `KL_UCB`

It does, as a child class of `UCB` (in `fit_predict_add_res()` any instance of `UCB` uses `pos_neg_train`)

File: replay/models/kl_ucb.py (hunk 1 above)
Comment (shashist) [reply replay/models/kl_ucb.py:10]: I would suggest to remove term `actor`. It is more likely to be used in Deep RL (quick googling "multi actor UCB" gives no result), also we don't use term actor in UCB code. 
For me without "actor" it looks clear, like `model treat log as non-personalized, e.g. item scores are the same for all users`.

Comment (Arkadiy-Vladimirov) [reply replay/models/kl_ucb.py:10]: ok, got it

File: replay/models/kl_ucb.py (hunk 2 above)
Comment (Arkadiy-Vladimirov) [reply replay/models/kl_ucb.py:116]: ok, no problem

**Review by shashist — APPROVED** [review]
Code part is ok. @Arkadiy-Vladimirov please sync this PR with recent changes and fix CI

**Comment by shashist** [comment]
Please sync with main and I will rerun CI

File: tests/models/test_ucb.py [hunk 3]
```diff
@@ -1,12 +1,15 @@
 # pylint: disable=redefined-outer-name, missing-function-docstring, unused-import
 import pytest
 
-from replay.models import UCB
-from tests.utils import create_dataset, log, log2, spark, sparkDataFrameEqual, sparkDataFrameNotEqual
-
-pyspark = pytest.importorskip("pyspark")
 from pyspark.sql import functions as sf
 
+<<<<<<< HEAD
```
Comment (bysheva) [inline tests/models/test_ucb.py:6]: Please clean up the merge issues

File: tests/models/test_ucb.py [hunk 4]
```diff
@@ -22,11 +25,10 @@ def log_ucb2(log2):
...
 
 
-@pytest.fixture
-def fitted_model(log_ucb):
-    model = UCB()
-    dataset = create_dataset(log_ucb)
-    model.fit(dataset)
+@pytest.fixture(params=[UCB(), KLUCB()])
+def fitted_model(log_ucb, model):
+>>>>>>> cfe340d (fixed tests,linter)
```
Comment (bysheva) [inline tests/models/test_ucb.py:30]: The same, merge issues

File: replay/models/__init__.py [hunk 5]
```diff
@@ -3,21 +3,31 @@
...
+    models with non-distributed training and distributed inference in PySpark.
 """
 
+from replay.models.admm_slim import ADMMSLIM
 from replay.models.als import ALSWrap
 from replay.models.association_rules import AssociationRulesItemRec
 from replay.models.base_rec import Recommender
-from replay.models.cat_pop_rec import CatPopRec
-from replay.models.cluster import ClusterRec
+from replay.models.base_torch_rec import TorchRecommender
```
Comment (bysheva) [inline replay/models/__init__.py:14]: Need to update imports list, the most of the changes are in the experimental submodule now

File: replay/models/kl_ucb.py [hunk 6]
```diff
@@ -0,0 +1,156 @@
+import math
+
+from typing import Optional
+from replay.models import UCB
+from pyspark.sql.types import DoubleType
```
Comment (bysheva) [inline replay/models/kl_ucb.py:5]: should be checked for pyspark availability, if PYSPARK_AVAILABLE...
Please see example here: https://github.com/sb-ai-lab/RePlay/blob/main/replay/models/knn.py

File: replay/models/kl_ucb.py [hunk 7]
```diff
@@ -0,0 +1,156 @@
...
+    log as non-personalized - item scores are same for all users.
+
+    If ``relevance`` column is not of 0/1 initially, then you have to decide
+    what kind of relevance has to be considered as positive and convert
+    ``relevance`` to binary format during preprocessing.
+
+    To provide a prediction, KL-UCB would sample a set of recommended items for
+    each user with probabilites proportional to obtained relevances.
+
+    >>> import pandas as pd
```
Comment (bysheva) [inline replay/models/kl_ucb.py:77]: Please check the example, now all models should support Dataset, please see doc example here https://github.com/sb-ai-lab/RePlay/blob/main/replay/models/ucb.py

**Review by bysheva — CHANGES_REQUESTED** [review]

**Comment by shashist** [comment]
Tests should be fixed. Currently they will result in `ModuleNotFoundError: No module named 'pyspark'` on gitlab (most recent) CI

**Pull request closed without merging**